From 284a89e4fcf17221d84b300fc5b1dc227e900719 Mon Sep 17 00:00:00 2001
From: Hwee Ren <hr@hweeren.com>
Date: Mon, 20 Apr 2026 15:14:45 +0800
Subject: [PATCH 11/11] updated Makefile, include riscv headers

---
 Makefile        | 2 ++
 user/lazytest.c | 1 +
 2 files changed, 3 insertions(+)

diff --git a/Makefile b/Makefile
index 4f604c0..505fa7c 100644
--- a/Makefile
+++ b/Makefile
@@ -144,8 +144,10 @@ UPROGS=\
 	$U/_zombie\
 	$U/_countvp_test\
 	$U/_countpp_test\
+	$U/_lazytest\
 	$U/_fault\
 
+
 fs.img: mkfs/mkfs README $(UPROGS)
 	mkfs/mkfs fs.img README $(UPROGS)
 
diff --git a/user/lazytest.c b/user/lazytest.c
index 6a9545a..1c365aa 100644
--- a/user/lazytest.c
+++ b/user/lazytest.c
@@ -1,4 +1,5 @@
 #include "kernel/types.h"
+#include "kernel/riscv.h"
 #include "kernel/stat.h"
 #include "user/user.h"
 
-- 
2.53.0

