From cea42dd22cebd9249465264fa0ad1009479caf5d Mon Sep 17 00:00:00 2001
From: Hwee Ren <hr@hweeren.com>
Date: Thu, 2 Apr 2026 00:47:26 +0800
Subject: [PATCH 4/4] changed MSG_SIZE to MSG_SIZE+1 to ensure string is null
 terminated

---
 user/race_test.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/user/race_test.c b/user/race_test.c
index dec7ecb..efa1ffa 100644
--- a/user/race_test.c
+++ b/user/race_test.c
@@ -87,7 +87,7 @@ void producer(const char* msg, struct buf_sem b) {
 
 void producer_consumer() {
     #define MSG_SIZE 48
-    const char msg[MSG_SIZE] = "cafefacefacefacefacefacefacefacefacecafecafecafe";
+    const char msg[MSG_SIZE +1] = "cafefacefacefacefacefacefacefacefacecafecafecafe";
     const int NUM_PROD = 2;
     int valid[256] = {0};
     for (const char* p = msg; *p != '\0'; p++) {
-- 
2.53.0

